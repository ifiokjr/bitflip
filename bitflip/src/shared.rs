pub struct AppError(anyhow::Error);

// Tell axum how to convert `AppError` into a response.
#[cfg(feature = "ssr")]
impl axum::response::IntoResponse for AppError {
	fn into_response(self) -> axum::response::Response {
		(
			http::StatusCode::INTERNAL_SERVER_ERROR,
			format!("App error: {}", self.0),
		)
			.into_response()
	}
}

// This enables using `?` on functions that return `Result<_, anyhow::Error>` to
// turn them into `Result<_, AppError>`. That way you don't need to do that
// manually.
impl<E> From<E> for AppError
where
	E: Into<anyhow::Error>,
{
	fn from(err: E) -> Self {
		Self(err.into())
	}
}
