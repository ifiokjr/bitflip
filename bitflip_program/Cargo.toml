[package]
name = "bitflip_program"
version = { workspace = true }
edition = { workspace = true }
homepage = { workspace = true }
license = { workspace = true }
repository = { workspace = true }
description = "The bitflip program for on-chain deployment"

[lib]
crate-type = ["cdylib", "lib"]
name = "bitflip_program"

[dependencies]
bitflags = { workspace = true }
bytemuck = { workspace = true }
const_str_to_pubkey = { workspace = true }
default-env = { workspace = true }
derive_more = { workspace = true, features = ["add", "add_assign"] }
fixed = { workspace = true }
log = { workspace = true, optional = true }
num_enum = { workspace = true }
serde = { workspace = true, features = ["derive"], optional = true }
serde-big-array = { workspace = true }
serde_with = { workspace = true, optional = true }
solana-package-metadata = { workspace = true }
solana-program = { workspace = true }
solana-sdk = { workspace = true, optional = true }
solana-security-txt = { workspace = true }
spl-associated-token-account = { workspace = true, features = ["no-entrypoint"] }
spl-pod = { workspace = true }
spl-token-2022 = { workspace = true, features = ["no-entrypoint"] }
spl-token-group-interface = { workspace = true }
spl-token-metadata-interface = { workspace = true }
spl-type-length-value = { workspace = true }
static_assertions = { workspace = true }
steel = { workspace = true, features = ["logs", "spl"] }
strum = { workspace = true, features = ["derive"], optional = true }
thiserror = { workspace = true }
typed-builder = { workspace = true, optional = true }
wasm_client_solana = { workspace = true, optional = true }

[dev-dependencies]
anyhow = { workspace = true }
assert2 = { workspace = true }
insta = { workspace = true, features = ["json", "redactions", "ron", "yaml"] }
rstest = { workspace = true }
serde_json = { workspace = true }
test-log = { workspace = true, features = ["log", "trace"] }

[features]
no-entrypoint = []
serde = [
	"dep:serde_with",
	"dep:serde",
	"spl-pod/serde-traits",
]
client = [
	"dep:log",
	"dep:solana-sdk",
	"dep:strum",
	"dep:typed-builder",
	"dep:wasm_client_solana",
	"serde",
]
ssr = ["wasm_client_solana?/ssr"]
js = [
	"no-entrypoint",
	"wasm_client_solana?/js",
]

[package.metadata.solana]
program-id = "5AuNvfV9Xi9gskJpW2qQJndQkFcwbWNV6fjaf2VvuEcM"
