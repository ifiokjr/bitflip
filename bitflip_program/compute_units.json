{
	"instructions": {
		"config_initialize": {
			"compute_units": 17569,
			"rounded_compute_units": 20000,
			"description": "Initialize the config"
		},
		"config_update_authority": {
			"compute_units": 2326,
			"rounded_compute_units": 10000,
			"description": "Update the authority of the config"
		},
		"flip_bit_test": {
			"compute_units": 64832,
			"rounded_compute_units": 70000,
			"description": "Flip a single bit from `0` to `1`"
		},
		"game_initialize": {
			"compute_units": 12979,
			"rounded_compute_units": 20000,
			"description": "Initialize the game"
		},
		"game_refresh_signer": {
			"compute_units": 2232,
			"rounded_compute_units": 10000,
			"description": "Refresh the signer"
		},
		"token_group_initialize": {
			"compute_units": 47429,
			"rounded_compute_units": 50000,
			"description": "Initialize the token group"
		},
		"token_initialize:bit": {
			"compute_units": 63184,
			"rounded_compute_units": 70000,
			"description": "Initialize group token: Bit"
		},
		"token_initialize:gibibit": {
			"compute_units": 59746,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: GibiBit"
		},
		"token_initialize:kibibit": {
			"compute_units": 59745,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: KibiBit"
		},
		"token_initialize:mebibit": {
			"compute_units": 59742,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: MebiBit"
		}
	},
	"version": "0.0.0"
}
