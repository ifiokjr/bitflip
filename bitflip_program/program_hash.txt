ebc22d77bc8ddd23f326598673e9fe7aa99fcdce9572fb8559a8d9782ade4ab8
