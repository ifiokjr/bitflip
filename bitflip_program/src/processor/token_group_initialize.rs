use solana_program::msg;
use steel::*;
use sysvar::rent::Rent;

use super::BitflipInstruction;
use crate::BitflipError;
use crate::ConfigState;
use crate::ID;
use crate::SEED_PREFIX;
use crate::SEED_TREASURY;
use crate::TokenMember;
use crate::cpi::token_group_initialize;
use crate::cpi::token_group_member_initialize;
use crate::create_pda_config;
use crate::create_pda_mint;
use crate::create_pda_treasury;

pub fn process_token_group_initialize(accounts: &[AccountInfo]) -> ProgramResult {
	use TokenMember::*;
	// load accounts
	let [
		authority_info,
		config_info,
		treasury_info,
		mint_bit_info,
		mint_kibibit_info,
		mint_mebibit_info,
		mint_gibibit_info,
		associated_token_program_info,
		token_program_info,
		system_program_info,
	] = accounts
	else {
		return Err(ProgramError::NotEnoughAccountKeys);
	};

	authority_info.is_signer()?.is_writable()?;
	config_info.is_type::<ConfigState>(&ID)?;
	treasury_info.has_owner(&system_program::ID)?;
	mint_bit_info.is_writable()?;
	mint_kibibit_info.is_writable()?;
	mint_mebibit_info.is_writable()?;
	mint_gibibit_info.is_writable()?;
	associated_token_program_info.is_program(&spl_associated_token_account::ID)?;
	token_program_info.is_program(&spl_token_2022::ID)?;
	system_program_info.is_program(&system_program::ID)?;

	let config = config_info.as_account::<ConfigState>(&ID)?;
	let config_key = create_pda_config(config.bump)?;
	let treasury_key = create_pda_treasury(config.treasury_bump)?;
	let mint_bit_key = create_pda_mint(Bit, Bit.bump(config))?;
	let mint_kibibit_key = create_pda_mint(Kibibit, Kibibit.bump(config))?;
	let mint_mebibit_key = create_pda_mint(Mebibit, Mebibit.bump(config))?;
	let mint_gibibit_key = create_pda_mint(Gibibit, Gibibit.bump(config))?;
	let treasury_seeds = &[SEED_PREFIX, SEED_TREASURY, &[config.treasury_bump]];

	if config_key.ne(config_info.key)
		|| treasury_key.ne(treasury_info.key)
		|| mint_bit_key.ne(mint_bit_info.key)
		|| mint_kibibit_key.ne(mint_kibibit_info.key)
		|| mint_mebibit_key.ne(mint_mebibit_info.key)
		|| mint_gibibit_key.ne(mint_gibibit_info.key)
	{
		return Err(ProgramError::InvalidSeeds);
	}

	if authority_info.key.ne(&config.authority) {
		return Err(BitflipError::Unauthorized.into());
	}

	let rent_sysvar = Rent::get()?;

	msg!("{}: initialize group", Bit.name());
	token_group_initialize(
		token_program_info,
		mint_bit_info,
		mint_bit_info,
		treasury_info,
		&[&treasury_seeds[..]],
	)?;

	for (info, member) in [
		(mint_kibibit_info, Kibibit),
		(mint_mebibit_info, Mebibit),
		(mint_gibibit_info, Gibibit),
	] {
		msg!("{}: initialize group member", member.name());
		token_group_member_initialize(
			token_program_info,
			info,
			info,
			treasury_info,
			mint_bit_info,
			treasury_info,
			&[treasury_seeds],
		)?;
	}

	for info in [
		mint_bit_info,
		mint_kibibit_info,
		mint_mebibit_info,
		mint_gibibit_info,
	] {
		let extra_lamports = rent_sysvar
			.minimum_balance(info.data_len())
			.checked_sub(info.lamports())
			.ok_or(ProgramError::ArithmeticOverflow)?;

		if extra_lamports > 0 {
			info.collect(extra_lamports, authority_info)?;
		}
	}

	Ok(())
}

#[repr(C)]
#[derive(Clone, Copy, Debug, Pod, Zeroable)]
pub struct TokenGroupInitialize {}

instruction!(BitflipInstruction, TokenGroupInitialize);

#[cfg(test)]
mod tests {
	use std::cell::RefCell;
	use std::rc::Rc;

	use assert2::check;
	use solana_sdk::bpf_loader_upgradeable;
	use solana_sdk::native_loader;

	use super::*;
	use crate::BitflipError;
	use crate::get_pda_config;
	use crate::get_pda_mint;
	use crate::get_pda_treasury;
	use crate::leak;

	#[test_log::test]
	fn validation_should_pass() -> anyhow::Result<()> {
		let accounts = create_account_infos();
		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::UnsupportedSysvar);

		Ok(())
	}

	#[test_log::test]
	fn should_have_enough_accounts() -> anyhow::Result<()> {
		let accounts = create_account_infos();
		let result = process_token_group_initialize(&accounts[0..7]);
		check!(result.unwrap_err() == ProgramError::NotEnoughAccountKeys);

		Ok(())
	}

	#[test_log::test]
	fn authority_should_be_signer() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let authority_info = &mut accounts[0];
		authority_info.is_signer = false;

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn authority_should_be_writable() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let authority_info = &mut accounts[0];
		authority_info.is_writable = false;

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn authority_should_be_from_config() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let authority_info = &mut accounts[0];
		authority_info.key = leak(Pubkey::new_unique());

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::Custom(BitflipError::Unauthorized.into()));

		Ok(())
	}

	#[test_log::test]
	fn config_should_be_pda() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let config_info = &mut accounts[1];
		config_info.key = leak(Pubkey::new_unique());

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidSeeds);

		Ok(())
	}

	#[test_log::test]
	fn config_should_have_valid_data() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let config_info = &mut accounts[1];
		config_info.data = Rc::new(RefCell::new(leak(vec![1u8; 8])));

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidAccountData);

		Ok(())
	}

	#[test_log::test]
	fn config_should_have_valid_owner() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let config_info = &mut accounts[1];
		config_info.owner = leak(Pubkey::new_unique());

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidAccountOwner);

		Ok(())
	}

	#[test_log::test]
	fn treasury_should_be_pda() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let treasury_info = &mut accounts[2];
		treasury_info.key = leak(Pubkey::new_unique());

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidSeeds);

		Ok(())
	}

	#[test_log::test]
	fn treasury_should_be_system_account() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let treasury_info = &mut accounts[2];
		treasury_info.owner = &ID;

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidAccountOwner);

		Ok(())
	}

	#[test_log::test]
	fn mint_bit_should_be_pda() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let mint_bit_info = &mut accounts[3];
		mint_bit_info.key = leak(Pubkey::new_unique());

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidSeeds);

		Ok(())
	}

	#[test_log::test]
	fn mint_bit_should_be_writable() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let mint_bit_info = &mut accounts[3];
		mint_bit_info.is_writable = false;

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn mint_kibibit_should_be_writable() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let mint_kibibit_info = &mut accounts[4];
		mint_kibibit_info.is_writable = false;

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn mint_kibibit_should_be_pda() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let mint_kibibit_info = &mut accounts[4];
		mint_kibibit_info.key = leak(Pubkey::new_unique());

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidSeeds);

		Ok(())
	}

	#[test_log::test]
	fn mint_mebibit_should_be_writable() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let mint_mebibit_info = &mut accounts[5];
		mint_mebibit_info.is_writable = false;

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn mint_mebibit_should_be_pda() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let mint_mebibit_info = &mut accounts[5];
		mint_mebibit_info.key = leak(Pubkey::new_unique());

		Ok(())
	}

	#[test_log::test]
	fn mint_gibibit_should_be_writable() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let mint_gibibit_info = &mut accounts[6];
		mint_gibibit_info.is_writable = false;

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn mint_gibibit_should_be_pda() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let mint_gibibit_info = &mut accounts[6];
		mint_gibibit_info.key = leak(Pubkey::new_unique());

		let result = process_token_group_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidSeeds);

		Ok(())
	}

	fn create_account_infos() -> [AccountInfo<'static>; 10] {
		let authority_key = leak(Pubkey::new_unique());
		let authority_lamports = leak(1_000_000_000);
		let authority_data = leak(vec![]);
		let config_key = leak(get_pda_config().0);
		let config_lamports = leak(0);
		let config_data = {
			let config_bump = get_pda_config().1;
			let mut data = vec![0u8; 8];
			let treasury_bump = get_pda_treasury().1;
			let mint_bump = get_pda_mint(TokenMember::Bit).1;
			let mint_kibibit_bump = get_pda_mint(TokenMember::Kibibit).1;
			let mint_mebibit_bump = get_pda_mint(TokenMember::Mebibit).1;
			let mint_gibibit_bump = get_pda_mint(TokenMember::Gibibit).1;
			data[0] = ConfigState::discriminator();
			data.append(
				&mut ConfigState::new(
					*authority_key,
					config_bump,
					treasury_bump,
					mint_bump,
					mint_kibibit_bump,
					mint_mebibit_bump,
					mint_gibibit_bump,
				)
				.to_bytes()
				.to_vec(),
			);
			leak(data)
		};
		let treasury_key = leak(get_pda_treasury().0);
		let treasury_lamports = leak(0);
		let treasury_data = leak(vec![]);
		let mint_bit_key = leak(get_pda_mint(TokenMember::Bit).0);
		let mint_bit_lamports = leak(0);
		let mint_bit_data = leak(vec![]);
		let mint_kibibit_key = leak(get_pda_mint(TokenMember::Kibibit).0);
		let mint_kibibit_lamports = leak(0);
		let mint_kibibit_data = leak(vec![]);
		let mint_mebibit_key = leak(get_pda_mint(TokenMember::Mebibit).0);
		let mint_mebibit_lamports = leak(0);
		let mint_mebibit_data = leak(vec![]);
		let mint_gibibit_key = leak(get_pda_mint(TokenMember::Gibibit).0);
		let mint_gibibit_lamports = leak(0);
		let mint_gibibit_data = leak(vec![]);
		let associated_token_program_key = leak(spl_associated_token_account::ID);
		let associated_token_program_lamports = leak(1_000_000_000);
		let associated_token_program_data = leak(vec![]);
		let token_program_key = leak(spl_token_2022::ID);
		let token_program_lamports = leak(1_000_000_000);
		let token_program_data = leak(vec![]);
		let system_program_lamports = leak(1_000_000_000);
		let system_program_data = leak(vec![]);

		let authority_info = AccountInfo::new(
			authority_key,
			true,
			true,
			authority_lamports,
			authority_data,
			&system_program::ID,
			false,
			u64::MAX,
		);
		let config_info = AccountInfo::new(
			config_key,
			false,
			true,
			config_lamports,
			config_data,
			&ID,
			false,
			u64::MAX,
		);
		let treasury_info = AccountInfo::new(
			treasury_key,
			false,
			true,
			treasury_lamports,
			treasury_data,
			&system_program::ID,
			false,
			u64::MAX,
		);
		let mint_bit_info = AccountInfo::new(
			mint_bit_key,
			false,
			true,
			mint_bit_lamports,
			mint_bit_data,
			&system_program::ID,
			false,
			u64::MAX,
		);
		let mint_kibibit_info = AccountInfo::new(
			mint_kibibit_key,
			false,
			true,
			mint_kibibit_lamports,
			mint_kibibit_data,
			&system_program::ID,
			false,
			u64::MAX,
		);
		let mint_mebibit_info = AccountInfo::new(
			mint_mebibit_key,
			false,
			true,
			mint_mebibit_lamports,
			mint_mebibit_data,
			&system_program::ID,
			false,
			u64::MAX,
		);
		let mint_gibibit_info = AccountInfo::new(
			mint_gibibit_key,
			false,
			true,
			mint_gibibit_lamports,
			mint_gibibit_data,
			&system_program::ID,
			false,
			u64::MAX,
		);
		let associated_token_program_info = AccountInfo::new(
			associated_token_program_key,
			false,
			false,
			associated_token_program_lamports,
			associated_token_program_data,
			&bpf_loader_upgradeable::ID,
			true,
			u64::MAX,
		);
		let token_program_info = AccountInfo::new(
			token_program_key,
			false,
			false,
			token_program_lamports,
			token_program_data,
			&bpf_loader_upgradeable::ID,
			true,
			u64::MAX,
		);
		let system_program_info = AccountInfo::new(
			&system_program::ID,
			false,
			false,
			system_program_lamports,
			system_program_data,
			&native_loader::ID,
			true,
			u64::MAX,
		);

		[
			authority_info,
			config_info,
			treasury_info,
			mint_bit_info,
			mint_kibibit_info,
			mint_mebibit_info,
			mint_gibibit_info,
			associated_token_program_info,
			token_program_info,
			system_program_info,
		]
	}
}
