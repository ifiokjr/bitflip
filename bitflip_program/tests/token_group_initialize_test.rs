#![cfg(feature = "client")]

use std::future::Future;

use assert2::check;
use bitflip_program::TOKEN_DECIMALS;
use bitflip_program::TokenMember;
use bitflip_program::get_pda_mint;
use bitflip_program::get_pda_treasury;
use bitflip_program::get_treasury_token_account;
use bitflip_program::token_group_initialize;
use shared::ToRpcClient;
use shared::create_authority_keypair;
use shared::create_config_accounts;
use shared::create_program_context_with_factory;
use shared::create_token_accounts;
use solana_sdk::compute_budget::ComputeBudgetInstruction;
use solana_sdk::transaction::VersionedTransaction;
use test_utils::create_insta_redaction;
use test_utils_solana::prelude::*;
use wasm_client_solana::solana_account_decoder::parse_account_data::SplTokenAdditionalData;
use wasm_client_solana::solana_account_decoder::parse_token::parse_token_v2;

mod shared;

#[ignore]
#[test_log::test(tokio::test)]
async fn token_group_initialize_test() -> anyhow::Result<()> {
	shared_token_group_initialize_test(create_banks_client_rpc).await?;
	Ok(())
}

#[cfg(feature = "test_validator")]
#[ignore]
#[test_log::test(tokio::test)]
async fn token_group_initialize_test_validator() -> anyhow::Result<()> {
	let compute_units = shared_token_group_initialize_test(create_validator_rpc).await?;
	let rounded_compute_units = bitflip_program::round_compute_units_up(compute_units);

	check!(rounded_compute_units <= 90_000);
	insta::assert_snapshot!(format!("{rounded_compute_units} CU"));
	shared::save_compute_units(
		"token_group_initialize",
		compute_units,
		"Initialize the token group",
	)?;

	Ok(())
}

async fn create_banks_client_rpc() -> anyhow::Result<impl ToRpcClient> {
	let provider = create_program_context_with_factory(|p| {
		let mut accounts = create_config_accounts();
		accounts.extend(create_token_accounts()?);

		for (key, account) in accounts {
			p.add_account(key, account.into());
		}

		Ok(())
	})
	.await?;
	Ok(provider)
}

#[cfg(feature = "test_validator")]
async fn create_validator_rpc() -> anyhow::Result<impl ToRpcClient> {
	let mut accounts = create_config_accounts();
	accounts.extend(create_token_accounts()?);

	let runner = shared::create_runner_with_accounts(accounts).await;

	Ok(runner)
}

async fn shared_token_group_initialize_test<
	T: ToRpcClient,
	Fut: Future<Output = anyhow::Result<T>>,
	Create: FnOnce() -> Fut,
>(
	create_provider: Create,
) -> anyhow::Result<u64> {
	let provider = create_provider().await?;
	let rpc = provider.to_rpc();
	let authority_keypair = create_authority_keypair();
	let authority = authority_keypair.pubkey();
	let treasury = get_pda_treasury().0;
	let treasury_bit_token_account = get_treasury_token_account(TokenMember::Bit);
	let recent_blockhash = rpc.get_latest_blockhash().await?;
	let instruction = token_group_initialize(&authority);
	let compute_limit_instruction = ComputeBudgetInstruction::set_compute_unit_limit(1_400_000);
	let mut transaction = VersionedTransaction::new_unsigned_v0(
		&authority,
		&[compute_limit_instruction, instruction],
		&[],
		recent_blockhash,
	)?;

	let simulation = rpc.simulate_transaction(&transaction).await?;
	log::info!("simulation: {simulation:#?}");

	let compute_units = simulation.value.units_consumed.unwrap();

	transaction.try_sign(&[&authority_keypair], None)?;

	let signature = rpc.send_and_confirm_transaction(&transaction).await?;
	rpc.confirm_transaction(&signature).await?;

	for ii in 0..4 {
		let member = TokenMember::try_from(ii)?;
		let mint = get_pda_mint(member).0;
		let treasury_redaction = create_insta_redaction(treasury, "treasury:pubkey");
		let mint_redaction = create_insta_redaction(mint, "mint:pubkey");
		let data = rpc.get_account_data(&mint).await?;
		let mint_account = parse_token_v2(
			&data,
			Some(&SplTokenAdditionalData::with_decimals(TOKEN_DECIMALS)),
		)?;

		log::info!("mint account: {mint_account:#?}");
		insta::assert_compact_json_snapshot!(mint_account, {
			".info.mintAuthority" => insta::dynamic_redaction(treasury_redaction.clone()),
			".info.freezeAuthority" => insta::dynamic_redaction(treasury_redaction.clone()),
			".info.extensions[0].state.authority" => insta::dynamic_redaction(treasury_redaction.clone()),
			".info.extensions[0].state.metadataAddress" =>	insta::dynamic_redaction(mint_redaction.clone()),
		  ".info.extensions[1].state.closeAuthority" => insta::dynamic_redaction(treasury_redaction.clone()),
			".info.extensions[2].state.authority" => insta::dynamic_redaction(treasury_redaction.clone()),
			".info.extensions[2].state.memberAddress" => insta::dynamic_redaction(mint_redaction.clone()),
			".info.extensions[2].state.groupAddress" => insta::dynamic_redaction(mint_redaction.clone()),
		  ".info.extensions[3].state.updateAuthority" => insta::dynamic_redaction(treasury_redaction.clone()),
		  ".info.extensions[3].state.mint" =>	insta::dynamic_redaction(mint_redaction.clone())
		});

		let data = rpc.get_account_data(&treasury_bit_token_account).await?;
		let parsed_treasury_bit_token_account = parse_token_v2(
			&data,
			Some(&SplTokenAdditionalData::with_decimals(TOKEN_DECIMALS)),
		)?;

		log::info!("token account: {parsed_treasury_bit_token_account:#?}");
		insta::assert_compact_json_snapshot!(parsed_treasury_bit_token_account, {
			".info.mint" => insta::dynamic_redaction(mint_redaction),
			".info.owner" => insta::dynamic_redaction(treasury_redaction.clone()),
			".info.closeAuthority" => insta::dynamic_redaction(treasury_redaction),
		});
	}
	Ok(compute_units)
}
