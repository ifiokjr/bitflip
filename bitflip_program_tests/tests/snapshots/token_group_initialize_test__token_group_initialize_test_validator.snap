---
source: bitflip_program/tests/token_group_initialize_test.rs
expression: "format!(\"{rounded_compute_units} CU\")"
snapshot_kind: text
---
50000 CU
